File: python/abstract_factory/abstract_factory.py
from abc import ABC


class IFurnitureFactory(ABC):
    def create_table(self):
        pass

    def create_chair(self):
        pass


class IChair(ABC):
    def sit_on(self):
        pass


class ArtChair(IChair):
    def sit_on(self):
        print("Sitting on art chair.")


class ITable(ABC):
    def eat_from(self):
        pass


class ArtTable(ITable):
    def eat_from(self):
        print("Eating from art table.")


class ArtFurnitureFactory(IFurnitureFactory):
    def create_table(self):
        return ArtTable()

    def create_chair(self):
        return ArtChair()


if __name__ == '__main__':
    factory = ArtFurnitureFactory()
    chair = factory.create_chair()
    table = factory.create_table()
    chair.sit_on()
    table.eat_from()

File: go/abstract_factory/abstract_factory.go
package main

import "fmt"

type IChair interface {
	SitOn()
}

type ITable interface {
	EatFrom()
}

type ArtChair struct {
}

func (receiver ArtChair) SitOn() {
	fmt.Println("Sitting on art chair.")
}

type ArtTable struct {
}

func (receiver ArtTable) EatFrom() {
	fmt.Println("Eating from art table.")
}

type IFurnitureFactory interface {
	CreateChair() IChair
	CreateTable() ITable
}

type ArtFurnitureFactory struct {
}

func (receiver ArtFurnitureFactory) CreateChair() IChair {
	return ArtChair{}
}

func (receiver ArtFurnitureFactory) CreateTable() ITable {
	return ArtTable{}
}

func main() {
	factory := ArtFurnitureFactory{}
	chair := factory.CreateChair()
	table := factory.CreateTable()
	chair.SitOn()
	table.EatFrom()
}
